internal/devconfig/config.go

```diff
@@ -104,19 +114,36 @@ func loadBytes(b []byte) (*Config, error) {
 }
 
 func (c *Config) LoadRecursive(lockfile *lock.File) error {
+	return c.loadRecursive(lockfile, map[string]bool{})
+}
+
+func (c *Config) loadRecursive(
+	lockfile *lock.File,
+	seen map[string]bool,
+) error {
 	included := make([]*Config, 0, len(c.Root.Include))
 
 	for _, includeRef := range c.Root.Include {
-		pluginConfig, err := plugin.LoadConfigFromInclude(includeRef, lockfile)
+		pluginConfig, err := plugin.LoadConfigFromInclude(
+			includeRef, lockfile, c.absRootPath)
 		if err != nil {
 			return errors.WithStack(err)
 		}
 
+		if seen[pluginConfig.Source.Hash()] {
+			return errors.Errorf("circular include detected: %s", includeRef)
+		}
+		seen[pluginConfig.Source.Hash()] = true
```

Comment:
For ease of debugging, one could improve the error message by showing the cyclical-path-of-includes. May need to track a bit more state...but feels doable.